sdks/python/apache_beam/io/tfrecordio.py

```diff
@@ -0,0 +1,274 @@
+
+    # Validate all length related payloads.
+    if len(buf) != buf_length_expected:
+      raise ValueError('Not a valid TFRecord. Fewer than %d bytes: %s' %
+                       (buf_length_expected, buf.encode('hex')))
+    length, length_mask_expected = struct.unpack('<QI', buf)
+    length_mask_actual = cls._masked_crc32c(buf[:8])
+    if length_mask_actual != length_mask_expected:
+      raise ValueError('Not a valid TFRecord. Mismatch of length mask: %s' %
+                       buf.encode('hex'))
+
+    # Validate all data related payloads.
+    buf_length_expected = length + 4
+    buf = file_handle.read(buf_length_expected)
+    if len(buf) != buf_length_expected:
+      raise ValueError('Not a valid TFRecord. Fewer than %d bytes: %s' %
+                       (buf_length_expected, buf.encode('hex')))
+    data, data_mask_expected = struct.unpack('<%dsI' % length, buf)
+    data_mask_actual = cls._masked_crc32c(data)
+    if data_mask_actual != data_mask_expected:
+      raise ValueError('Not a valid TFRecord. Mismatch of data mask: %s' %
+                       buf.encode('hex'))
+
+    # All validation checks passed.
+    return data
+
+
+class _TFRecordSource(filebasedsource.FileBasedSource):
+  """A File source for reading files of TFRecords.
+
+  For detailed TFRecords format description see:
+    https://www.tensorflow.org/versions/master/api_docs/python/python_io.html#tfrecords-format-details
+  """
+
+  def __init__(self,
+               file_pattern,
+               coder,
+               compression_type):
+    """Initialize a TFRecordSource.  See ReadFromTFRecord for details."""
+    super(_TFRecordSource, self).__init__(
+        file_pattern=file_pattern,
+        compression_type=compression_type,
+        splittable=False)
+    self._coder = coder
+
+  def read_records(self, file_name, offset_range_tracker):
+    if offset_range_tracker.start_position():
+      raise ValueError('Start position not 0:%s' %
+                       offset_range_tracker.start_position())
+
+    current_offset = offset_range_tracker.start_position()
+    with self.open_file(file_name) as file_handle:
+      while True:
+        if not offset_range_tracker.try_claim(current_offset):
+          raise RuntimeError('Unable to claim position: %s' % current_offset)
+        record = _TFRecordUtil.read_record(file_handle)
+        if record is None:
+          return  # Reached EOF
+        else:
+          current_offset += _TFRecordUtil.encoded_num_bytes(record)
+          yield self._coder.decode(record)
+
+
+class ReadFromTFRecord(PTransform):
+  """Transform for reading TFRecord sources."""
+
+  def __init__(self,
+               file_pattern,
+               coder=coders.BytesCoder(),
+               compression_type=fileio.CompressionTypes.AUTO,
+               **kwargs):
+    """Initialize a ReadFromTFRecord transform.
+
+    Args:
+      file_pattern: The file pattern to read from as a local file path or a GCS
```

Comment:
file_pattern is a file glob to read. It can be local, GCS or anything else FileBasedSource can open. I do not think we should specify GCS explicitly.